### PR TITLE
Update cors configuration

### PR DESCRIPTION
**Pull request by manti-by**
Decription https://github.com/OpenUpSA/muni-portal-backend/issues/41

### CODE REVIEW

File: muni_portal/settings.py [hunk 1]
```diff
@@ -198,8 +199,16 @@
 WAGTAILAPI_BASE_URL = env.str("WAGTAILAPI_BASE_URL", None)
 FRONTEND_BASE_URL = env.str("FRONTEND_BASE_URL", None)
 
-CORS_URLS_REGEX = r"^/api/.*$"
-CORS_ALLOW_ALL_ORIGINS = True
+CORS_URLS_REGEX = r"^/api/(accounts|token)/.*$"
+CORS_ALLOWED_ORIGIN_REGEXES = [re.compile(x) for x in env.list("CORS_ALLOWED_ORIGIN_REGEXES", [])]
+CORS_ALLOW_HEADERS = [
+    "accept",
```
Comment (jbothma) [inline muni_portal/settings.py:205]: please use corsheaders.defaults.default_headers as per the docs instead of hardcoding these

File: README.md [hunk 2]
```diff
@@ -129,4 +129,14 @@ Web Push Notifications
...
+
+
+Security
+--------
+
+- CORS_URLS_REGEX - Currently allows access to user account API. [Example here](https://regex101.com/r/Ui3hn2/3).
+
+- CORS_ALLOWED_ORIGIN_REGEXES - Should be configured via environment. [Example here](https://regex101.com/r/q6jWFA/2/).
+
+- List of allowed headers: Accept, Accept-Encoding, Content-Type, Charset, HTTP_AUTHORIZATION, Origin.
```
Comment (jbothma) [inline README.md:142]:
```suggestion
Security
--------

Exploits to watch out for with security of this service (not comprehensive):

- Do not allow credentialed access to /admin or the Webflow API where an admin might be logged in and an attacker uses their session to fetch or modify information (CORS and CSRF)
  - Do not allow cookie or basic authentication to user-endpoints like fetching/updating their profile or performing actions on their behalf (CORS and CSRF)
    - By only allowing token-based authentication, the browser can not be tricked into supplying credentials
    - CORS by default disallows the browser from automatically supplying credentials - be careful about changing this behaviour


- CORS_URLS_REGEX - Currently allows access to user account API. [Example here](https://regex101.com/r/Ui3hn2/3).

- CORS_ALLOWED_ORIGIN_REGEXES - Should be configured via environment. [Example here](https://regex101.com/r/q6jWFA/2/).

- List of allowed headers: Accept, Accept-Encoding, Content-Type, Charset, HTTP_AUTHORIZATION, Origin.
```

File: muni_portal/settings.py [hunk 3]
```diff
@@ -198,8 +200,9 @@
 WAGTAILAPI_BASE_URL = env.str("WAGTAILAPI_BASE_URL", None)
 FRONTEND_BASE_URL = env.str("FRONTEND_BASE_URL", None)
 
-CORS_URLS_REGEX = r"^/api/.*$"
-CORS_ALLOW_ALL_ORIGINS = True
+CORS_URLS_REGEX = r"^/api/(accounts|token)/.*$"
+CORS_ALLOWED_ORIGIN_REGEXES = [re.compile(x) for x in env.list("CORS_ALLOWED_ORIGIN_REGEXES", [])]
+CORS_ALLOW_HEADERS = default_headers
```
Comment (jbothma) [inline muni_portal/settings.py:205]: but remember we have to add the custom header to this! that was the point of this task :) so it's default + ...

Comment (manti-by) [reply muni_portal/settings.py:205]: sorry, forgot

File: README.md [hunk 4]
```diff
@@ -129,4 +129,22 @@ Web Push Notifications
...
+Security
+--------
+
+Exploits to watch out for with security of this service (not comprehensive):
+
+- Do not allow credentialed access to /admin or the Webflow API where an admin might be logged in and an attacker uses their session to fetch or modify information (CORS and CSRF)
+  - Do not allow cookie or basic authentication to user-endpoints like fetching/updating their profile or performing actions on their behalf (CORS and CSRF)
+    - By only allowing token-based authentication, the browser can not be tricked into supplying credentials
+    - CORS by default disallows the browser from automatically supplying credentials - be careful about changing this behaviour
+
```
Comment (jbothma) [inline README.md:144]:
```suggestion
Important configuration for security
```

File: README.md [hunk 5]
```diff
@@ -129,4 +129,22 @@ Web Push Notifications
...
+Security
+--------
+
+Exploits to watch out for with security of this service (not comprehensive):
+
+- Do not allow credentialed access to /admin or the Webflow API where an admin might be logged in and an attacker uses their session to fetch or modify information (CORS and CSRF)
+  - Do not allow cookie or basic authentication to user-endpoints like fetching/updating their profile or performing actions on their behalf (CORS and CSRF)
+    - By only allowing token-based authentication, the browser can not be tricked into supplying credentials
+    - CORS by default disallows the browser from automatically supplying credentials - be careful about changing this behaviour
+Important configuration for security
```
Comment (jbothma) [inline README.md:144]:
```suggestion
Specific approach to CORS and CSRF

- the resources relevant to CSRF are the ones protected by automated credential headers like session cookie or basic authentication.
- our custom resources disallow both those, and are therefore secure against CSRF
- wagtail (API and CMS) and the django admin uses session cookies
- any resources that work with automated credential headers like cookies or basic auth must be protected from access by unsafe origins using CORS.
- CORS must therefore not be allowed for admin or wagtail resources.
- CORS may only be allowed for our custom resources that only allow token authentication

Important configuration for security
```

File: README.md [hunk 6]
```diff
@@ -129,4 +129,31 @@ Web Push Notifications
...
+Security
+--------
+
+Exploits to watch out for with security of this service (not comprehensive):
+
+- Do not allow credentialed access to /admin or the Webflow API where an admin might be logged in and an attacker uses their session to fetch or modify information (CORS and CSRF)
+  - Do not allow cookie or basic authentication to user-endpoints like fetching/updating their profile or performing actions on their behalf (CORS and CSRF)
+    - By only allowing token-based authentication, the browser can not be tricked into supplying credentials
+    - CORS by default disallows the browser from automatically supplying credentials - be careful about changing this behaviour
+Specific approach to CORS and CSRF
```
Comment (jbothma) [inline README.md:144]:
```suggestion

Specific approach to CORS and CSRF
```

**Pull request merged**